pulp_rpm/tests/functional/api/test_sync.py

```diff
@@ -75,1390 +66,988 @@
-        if sync_policy == "mirror_complete":
-            self.assertEqual(publications_api.list().count, 1)
-            self.assertEqual(len(task.created_resources), 2)
-        else:
-            self.assertEqual(len(task.created_resources), 1)
+
+
+@pytest.mark.parallel
+def test_core_metadata(init_and_sync, rpm_package_api):
+    """Test that the metadata returned by the Pulp API post-sync matches what we expect.
+
+    Do the following:
+
+    1. Sync a repo.
+    2. Query package metadata from the API.
+    3. Match it against the metadata that we expect to be there.
+    """
+    repository, _ = init_and_sync(url=RPM_COMPLEX_FIXTURE_URL, policy="on_demand")
+
+    package = rpm_package_api.list(
+        name=RPM_COMPLEX_PACKAGE_DATA["name"], repository_version=repository.latest_version_href
+    ).results[0]
+    package = package.to_dict()
+    # delete pulp-specific metadata
+    package.pop("pulp_href")
+    package.pop("pulp_created")
+
+    # sort file and changelog metadata
+    package["changelogs"].sort(reverse=True)
+    for metadata in [package, RPM_COMPLEX_PACKAGE_DATA]:
+        # the list-of-lists can't be sorted easily so we produce a string representation
+        files = []
+        for f in metadata["files"]:
+            files.append(
+                "{basename}{filename} type={type}".format(
+                    basename=f[1], filename=f[2], type=f[0] or "file"
+                )
+            )
+        metadata["files"] = sorted(files)
+
+    assert all(
+        value == RPM_COMPLEX_PACKAGE_DATA[key]
+        for key, value in package.items()
+        if key != "time_file"
+    )
+
+
+@pytest.mark.parallel
+def test_treeinfo_metadata(init_and_sync, rpm_content_distribution_trees_api):
+    """Test that the metadata returned by the Pulp API post-sync matches what we expect.
+
+    Do the following:
+
+    1. Sync a repo.
+    2. Query treeinfo metadata from the API.
+    3. Match it against the metadata that we expect to be there.
+    """
+    repository, _ = init_and_sync(url=RPM_KICKSTART_FIXTURE_URL, policy="on_demand")
+
+    distribution_tree = rpm_content_distribution_trees_api.list(
+        repository_version=repository.latest_version_href
+    ).results[0]
+    distribution_tree = distribution_tree.to_dict()
+    # delete pulp-specific metadata
+    distribution_tree.pop("pulp_href")
+
+    # sort kickstart metadata so that we can compare the dicts properly
+    for d in [distribution_tree, RPM_KICKSTART_DATA]:
+        d["addons"] = sorted(d["addons"], key=lambda x: x["addon_id"])
+        d["images"] = sorted(d["images"], key=lambda x: x["path"])
+        d["checksums"] = sorted(d["checksums"], key=lambda x: x["path"])
+        d["variants"] = sorted(d["variants"], key=lambda x: x["variant_id"])
+
+    for image in distribution_tree["images"]:
+        image.pop("artifact")
+
+    assert distribution_tree == RPM_KICKSTART_DATA
```

Comment:
Likewise ^

Reply:
Re: https://github.com/pulp/pulp_rpm/pull/2922#discussion_r1064156846.

The assertion difference should be shown more clearly between non-matching dictionaries in pytest. In unittest, we had to explicitly declare what needs to be shown after the failure. Am I right, @bmbouter, @dkliban?